### DISCUSSION TITLE
Using the Plastic Deformation Gradient (Fp) Crystal Plasticity

### DISCUSSION THREAD

**Discussion by Vandenbg** [idaholab/moose#21950] [Q&A General] [answered]
Hello, I am attempting to use the Plastic Deformation Gradient from the governing equations of [Crystal Plasticity Stress](https://mooseframework.inl.gov/source/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.html) to calculate the dislocation density in the material.

I am unsure if this is possible, so far I have found that it may be possible to output the deformation gradient using output_properties. This would require knowing the name which the deformation gradient is tied to within the file, is there a way to find out the possible material properties that can be output using that line? From here I would want to use the gradient as a variable so I could apply a function to it. 

Other than that, I have thought about using post processors or auxvariables, however, I am unsure of how to connect the deformation gradient to these. 

Any advice/ direction would be of great assistance.

**Comment by GiudGiud** [comment]
Hello

the deformation is stored in the disp_xyz variables, the gradient should be the strain tensor material property, named as strain_xy for the relevant component. You can trigger material property output to get it or you can add the relevant Material...Aux to output it

Guillaume

**Reply by Vandenbg** [reply]
Thank you for the reply. This is mostly what I am looking for, however, I need the components of the strain tensor, plastic and elastic. In the aforementioned syntax page, the total, elastic, and plastic tensors are all used for different calculations. Are these stored in two different variables? 

As for variable naming conventions you explained, is there a page with details I can reference?

**Comment by sapitts** [comment, marked as answer]
Hi @Vandenbg ,

In the crystal plasticity classes, we define the plastic deformation gradient and the total deformation gradient with respect to the reference configuration. @GiudGiud's answer refers to the tensor mechanics formulation derived with respect to the current configuration (in which the strain is Almansi strain calculated from the left Cauchy-Green deformation tensor).

Since you mention wanting to calculate dislocation density values, I suspect you will want the right Cauchy-Green plastic deformation tensor. In the ComputeMultipleCrystalPlasticityStress class, the plastic deformation gradient is given the material property name `_plastic_deformation_gradient`, see https://github.com/idaholab/moose/blob/712a86c3518eef66e5e36cfcf02c55025eed85be/modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.C#L82

What type of dislocation density are you trying to calculate?

**Reply by Vandenbg** [reply]
Thank you for the clarifications. I am trying to calculate the number of dislocations at each point in my mesh as it is deformed via nano-indentation. To do so, I am following [this reference](https://github.com/idaholab/moose/files/9447553/1252189.pdf) which details a method to calculate the dislocation density tensor. In this reference, page 3, section 2, equation 3 is the route I am pursuing. If another method for calculating the number of dislocations and their position is available within MOOSE, I am open to suggestions. 

I appreciate your insight.

**Reply by sapitts** [reply]
Ah, the Nye's tensor calculation for the geometrically necessary dislocations. You'll probably want to calculate that quantity as a material property, although material properties are computed at individual quadrature points (the `_qp` you see in the crystal plasticity code). A Nye's tensor calculated directly within a crystal plasticity model will depend on the plastic deformation gradient at just that single qp and will introduce mesh-dependence. Is the mesh-dependence issue why you asked about output options?

**Reply by Vandenbg** [reply]
Unfortunately it is not, it just happened to be the first step I identified. While I'd like to do the calculation within MOOSE, It might not be possible and a separate script for data analysis may be needed. I planned to use a series of functions along side the deformation gradient to calculate the dislocation density at each point. I believe [functions in MOOSE](https://mooseframework.inl.gov/syntax/Functions/index.html) have the capabilities for the calculation. How do calculations work for material properties, can I still use functions? I am unsure how mesh dependance will play-out, I have not put thought into that issue.

**Reply by ngrilli** [reply]
@Vandenbg you may want to calculate the geometrically necessary dislocation density starting from Fp 
by using a minimization procedure as described here:
https://www.sciencedirect.com/science/article/pii/S0749641918300068
This can be done in MOOSE or as postprocessing and I may have some python script to share for this.

Another perspective is to use a dislocation-based constitutive model in which the time evolution of the GND is calculated
during the simulation based on the gradients of the slip rate on the different slip systems, as done here:
https://www.sciencedirect.com/science/article/pii/S0749641920304940
see equations (48) and (49).

I have implemented this second option here:
https://github.com/ngrilli/c_pfor_am/blob/main/test/tests/SlipGradients/polycrystal_initial_GND.i

The advantage is that it does not require a specific assumption
about the energy minimization to discriminate between multiple solutions (i.e. given Fp, the GND densities are not unique) 
and gives a unique solution for the GND density on each slip system.

Trust this helps,
Nicolò

**Reply by Vandenbg** [reply]
Nicolò, Stephanie, I appreciate your insights on this issue. I'll pursue the second method given your advice.

**Reply by sapitts** [reply]
Thanks for sharing the Demir paper @ngrilli!
